### ISSUE TITLE
movementX/Y when moving out of or into an iframe

### ISSUE THREAD

**Issue by NavidZ** [w3c/pointerlock#16]
@scheib @RByers
What is the definition of these two properties when the pointer is moving across boundaries of the iframes? As soon as mouse enters another iframe it is like it is out of the whole window from the perspective of the other iframe. So if it re-enters an iframe from another one are the movementX/Y attributes supposed to be zero as if the previous position of the pointer was not known?

**Comment by scheib** [comment]
The spec requires zeros at window boundaries because the user agent can't know what the relative motion of the pointer was. However, it does know this when crossing iframe boundaries and I think we should provide the movementX/Y values. The spec doesn't address this specifically, but I think implicitly allows this fine.

However, let me know if in the code you see this to not be true, or if there's something else I'm missing.

The actual goal is that when entering iframes the movementX/Y values are usable and don't represent a large jump in value. E.g exiting an iframe at the top center, and then re-entering slowly from the right edge should show only a small leftward motion in movementX.

**Comment by NavidZ** [comment]
Right now in Chrome, from the iframe point of view the first event it gets has movementX/Y set as the difference between the current event and the last event. So if the last event (happened outside of this iframe) is far apart it does see a jump event though the first event that was delivered to this iframe might be near the boundary.

**Comment by scheib** [comment]
;( Ok, well that's not desired, so we can either see if the impl can work
around that, or I'm also OK adding to the spec a clearing at iframe
boundaries just to make impls easy... I don't think it would have a large
negative impact to applications to miss that first data packet, while the
large jumps are very seriously broken.

**Comment by mustaqahmed** [comment]
I think from a user's perspective, this is similar to the case when a locked mouse pointer (in a page w/o subframes) leaves the window & comes back:
- If the page sees the mousemoves outside its window (through OS level locking support), why an iframe won't?
- If the page doesn't see the mousemoves outside its window, the only feasible solution seems to be to compute the offsets from the last event the page knows about when the mouse comes back. The iframe case seems no different.

Assuming I didn't miss anything, can't the spec simply say that the "last mouse event" used in offset calculation is "last in current browsing context" (leaving the "current browsing context" defined in OS-dependent way)?

**Comment by RByers** [comment]
>> So if the last event (happened outside of this iframe) is far apart it does see a jump event though the first event that was delivered to this iframe might be near the boundary.

> Ok, well that's not desired,

I thought this was working as intended.  What I believe Navid is saying is that the event inside the new iframe includes all movement that occurred after the last event delivered to the previous frame (still inside the window) but never any more.  So this includes some travel that wasn't over the target frame, but still not a "big jump" with magnitude greater than that in other (non-iframe) cases.  This is exactly what the spec says, and I think what we want.

It was only his initial implementation for the touchscreen case where this wasn't working correctly IMHO.

**Comment by RByers** [comment]
> If the page doesn't see the mousemoves outside its window, the only feasible solution seems to be to compute the offsets from the last event the page knows about when the mouse comes back. The iframe case seems no different.

I think the Iframe case is fundamentally different from the window case for the following reason:
- `mouseleave` events are not sent to the outer frame when the mouse moves into the boundaries of a containing frame.  So from the perspective of the code in the outer frame, the mouse hasn't "left" at all.  This was @NavidZ's argument to me when I asked him this question.
- Iframes can be used when composing an application in a way transparent to the user.  It's often not part of the user's mental model of switching applications when the mouse moves between frames.

**Comment by NavidZ** [comment]
> I thought this was working as intended. What I believe Navid is saying is that the event inside the new iframe includes all movement that occurred after the last event delivered to the previous frame (still inside the window) but never any more. So this includes some travel that wasn't over the target frame, but still not a "big jump" with magnitude greater than that in other (non-iframe) cases. This is exactly what the spec says, and I think what we want.
> 
> It was only his initial implementation for the touchscreen case where this wasn't working correctly IMHO.

Forget about my original implementation. I raised this issue when I was trying to write a [test](https://github.com/w3c/web-platform-tests/pull/4571) (for the new implementation) and realized I really cannot rely on the sum of all movementX of the events received by the inner iframe to be the width of the frame when pointer spans the whole iframe. It could be any number less or greater than the width depending on how fast/slow I go in or out of the iframe. That is why I put that <10 magnitude for that test. The jump is definitely not more that what the outer frame would see if there were no innerframe at all. But I was wondering if even that jump is desired for the innerframe or not.

**Comment by scheib** [comment]
I agree that a pure horizontal movement across an iframe is not certain to have sum(movementX) == iframe width. The concern would be if the movement values were computed from the last pointer position over the iframe, instead of the last one received by the OS. 

Thanks for the clarification RByers, I think we're on the same page, but let me diagram this out so we're really sure:

Here's a diagram of a mouse moving from left to right across a browser window two times. The mouse events received by the OS are denoted by alphabet letters a-z. 

```
         +------------------------------------------------------+
         | Browser Window Title                                 |
         +------------------------------------------------------+
         |                                                      |
         |                an iframe in the web contents:        |
         |                .........................             |
         |                :                       :             |
  a      |      b         : c          d          :  e        f |     g  
         |                :                       :             |          h
     t   |     u          :   v            w      :   x     y   | z   
 s       |                :                       :             |            i
         |                :.......................:             |
  r      +------------------------------------------------------+          j
       q                                                              k 
                 p           o          n          m          l    
```

Mouse events a, g-t, and z are received by the OS from the pointing device, but NOT by the browser. The browser will not be able to use them to compute deltas for movementX/Y.

Event b must have movementX/Y == 0, because we are unable to compute the delta from a to b.

Event u must have movementX/Y == 0, avoiding the easy to implement error of measuring the delta from f to u (that happens if we just compute delta from last mouse data received by the browser).

Event c would ideally have movementX/Y set by the delta from b to c. But, I'm willing to also spec this as == 0 if there is some implementation issue that arrises.

Ditto for events e, v, x.

**Comment by RByers** [comment]
Thanks @scheib, that makes it very concrete and I agree completely.  I don't see any good reason why event c would need to have movementX/Y==0 so I'd prefer we leave things as spec'd.

I think the only challenge here is that @NavidZ was trying to test events c and v in relative isolation from the other events.  I don't think that makes sense.  The only way to validate that the movement in events c and v are correct is to communicate with the parent frame somehow (so one frame or the other knows about the b-c and u-v delta).

**Comment by scheib** [comment]
Closing this issue without spec changes, as @RByers indicates in previous comment https://github.com/w3c/pointerlock/issues/16#issuecomment-275151076 and with no contradiction raised after.

**Comment by NavidZ** [comment]
cc @eirage
I wanted to revive this issue and also bring up the point regarding the OOPIF iframes. I do believe a better and symmetric design for an embedded iframe would be to get the movementX/Y of the first event when we move in and out of an iframe. This way no matter when the iframe is embedded or it is the main frame it sees the same behavior. @smaug---- do you have any opinion about this?
